File: downloads/code/LeetCode/Java/Product-of-Array-Except-Self.java
public class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ret = new int[n];
        //累乘nums[i]左面的
        ret[0] = 1;
        for (int i=1;i<n;i++){
            ret[i] = nums[i-1]*ret[i-1]; 
        }
        
        // 类成nums[i]右边的
        int right = 1;
        for (int i=n-1;i>=0;i--){
            ret[i] *= right;
            right *= nums[i];
        }
        return ret;
    }
}

File: downloads/code/LeetCode/Python/Product-of-Array-Except-Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        ret = [0]*len(nums)
        if not nums:
            return ret
        product = 1
        
        n = 0
        for i in nums:
            if i == 0:
                n += 1
            else:
                product *= i
        # if nums contains more than two "0", return 0
        if n>1:
            return ret

        for i in range(len(nums)):
            if nums[i] == 0:
                ret[i] = product
            else:
                # if nums have one "0"
                if n == 1:
                    ret[i] = 0
                else:
                    ret[i] = product/nums[i]
        return ret
